src/control/positionControl/BBTrajectories/WorldObjects.cpp

```diff
@@ -0,0 +1,183 @@
+        if (!canEnterDefenseArea(robotId)) {
+            for (int i = 0; i<pathPoints.size(); i++) {
+                if (rtt::ai::FieldComputations::pointIsInDefenseArea(*field, pathPoints[i], true, 0) ||
+                    rtt::ai::FieldComputations::pointIsInDefenseArea(*field, pathPoints[i], false,
+                                                                     0.2 + rtt::ai::Constants::ROBOT_RADIUS())) {
+                    collisionDatas.emplace_back(CollisionData{pathPoints[i], pathPoints[i]});
+                    collisionTimes.emplace_back(i * timeStep);
+                    return;
+                }
+            }
+        }
+    }
+
+    void WorldObjects::calculateBallCollisions(std::vector<CollisionData> &collisionDatas, std::vector<double> &collisionTimes,
+                                               std::vector<Vector2> pathPoints, double timeStep) {
+        if (ruleset.minDistanceToBall > 0) {
+            auto startPositionBall = world->getWorld()->getBall()->get()->getPos();
+            auto VelocityBall = world->getWorld()->getBall()->get()->getFilteredVelocity();
+            std::vector<Vector2> ballTrajectory;
+
+            //TODO: improve ball trajectory approximation
+            //Current approximation assumes it continues on the same path with the same velocity, and we check 1 second deep
+            double time = 0;
+            double ballAvoidanceTime = 1;
+            while (pathPoints.size() * timeStep > time && time < ballAvoidanceTime) {
+                ballTrajectory.emplace_back(startPositionBall + VelocityBall * time);
+                time += timeStep;
+            }
+
+            // Check each timeStep for a collision with the ball, or during ball placement if its too close to the 'ballTube'
+            auto ballTube = LineSegment(startPositionBall, rtt::ai::GameStateManager::getRefereeDesignatedPosition());
+            for (int i = 0; i < ballTrajectory.size(); i++) {
+                if (ruleset.minDistanceToBall > (pathPoints[i] - ballTrajectory[i]).length()
+                    || (gameState.getStrategyName() == "ball_placement_them"
+                        && ruleset.minDistanceToBall > ballTube.distanceToLine(pathPoints[i]))) {
+                    collisionDatas.emplace_back(CollisionData{ballTrajectory[i], pathPoints[i]});
+                    collisionTimes.emplace_back(i * timeStep);
+                    return;
+                }
+            }
+        }
+    }
+
+    void
+    WorldObjects::calculateEnemyRobotCollisions(rtt::BB::BBTrajectory2D BBTrajectory, std::vector<CollisionData> &collisionDatas,
+                                                std::vector<double> &collisionTimes, const std::vector<Vector2> &pathPoints,
+                                                double timeStep) {
+        auto theirRobots = world->getWorld()->getThem();
+
+        for (int j = 0; j < pathPoints.size(); j++) {
+            for (int i = 0; i < theirRobots.size() - 1; i++) {
```

Comment:
Typically, people will reverse the i and j here, using i for the outer loop and j for the inner loop.